applications/StructuralMechanicsApplication/custom_constitutive/linear_j2_plasticity_3d.cpp

```diff
@@ -0,0 +1,471 @@
+}
+
+//************************************************************************************
+//************************************************************************************
+
+double& LinearJ2Plasticity3D::GetValue(const Variable<double>& rThisVariable,
+                                       double& rValue)
+{
+    if(rThisVariable == INELASTIC_FLAG){
+        rValue = mInelasticFlag;
+    }
+
+    return rValue;
+}
+
+void LinearJ2Plasticity3D::InitializeMaterial(const Properties& material_prop,
+                                              const GeometryType& rElementGeometry,
+                                              const Vector& rShapeFunctionsValues)
+{
+    mPlasticStrainOld = ZeroVector(this->GetStrainSize());
+    mAccumulatedPlasticStrainOld = 0.0;
+}
+
+void LinearJ2Plasticity3D::FinalizeSolutionStep(
+    const Properties& rMaterialProperties,
+    const GeometryType& rElementGeometry,
+    const Vector& rShapeFunctionsValues,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    mPlasticStrainOld = mPlasticStrain;
+    mAccumulatedPlasticStrainOld = mAccumulatedPlasticStrain;
+}
+
+void LinearJ2Plasticity3D::CalculateMaterialResponsePK1(Parameters& rValues)
+{
+    CalculateMaterialResponseCauchy(rValues);
+}
+
+void LinearJ2Plasticity3D::CalculateMaterialResponsePK2(Parameters& rValues)
+{
+    CalculateMaterialResponseCauchy(rValues);
+}
+
+void LinearJ2Plasticity3D::CalculateMaterialResponseKirchhoff(Parameters& rValues)
+{
+    CalculateMaterialResponseCauchy(rValues);
+}
+
+void LinearJ2Plasticity3D::CalculateMaterialResponseCauchy(Parameters& rValues)
+{
+    Flags &Options = rValues.GetOptions();
+
+    const Properties& rMaterialProperties = rValues.GetMaterialProperties();
+    Vector& strain_vector = rValues.GetStrainVector();
+    Vector& stress_vector = rValues.GetStressVector();
+
+    if (Options.Is(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR)) {
+        Matrix& ConstitutiveMatrix = rValues.GetConstitutiveMatrix();
+        CalculateElasticMatrix(ConstitutiveMatrix, rMaterialProperties);
+    }
+
+    if (Options.Is(ConstitutiveLaw::COMPUTE_STRESS)) {
+        if (rValues.GetProcessInfo().Has(INITIAL_STRAIN)) {
+            noalias(strain_vector) += rValues.GetProcessInfo()[INITIAL_STRAIN];
+        }
+        Matrix elastic_tensor;
+        Matrix& tangent_tensor = rValues.GetConstitutiveMatrix();
+        const double hardening_modulus = rMaterialProperties[ISOTROPIC_HARDENING_MODULUS];
+        const double delta_k = rMaterialProperties[INFINITY_HARDENING_MODULUS];
+        const double hardening_exponent = rMaterialProperties[HARDENING_EXPONENT];
+        const double E = rMaterialProperties[YOUNG_MODULUS];
+        const double poisson_ratio = rMaterialProperties[POISSON_RATIO];
+        const double mu = E / (2. + 2. * poisson_ratio);
+        const double volumetric_modulus = E / (3. * (1. - 2. * poisson_ratio));
+        const double sqrt_two_thirds = std::sqrt(2.0 / 3.0); // =0.8164965809277260
```

Comment:
i'll accept this, but just as a hint you coud make this sort of thing as consexpr so to guarantee they are evaluated at compile time